ramips: Add support for 6.12 kernel version

namiltd opened the pull request:
Kernel 6.12 update of target ramips.

Tested on:
- mt7620 - TP-Link Archer C5 v4 (rt2800soc + mt7612) - WiFi performance degradation
- mt7620 - Phicomm K2 (rt2800soc + mt7612)
- mt7621 - Xiaomi Mi Router 3G (mt7603 + mt7612)  
- mt7621 - Linksys EA6350 v4 (mt7603 + mt7613)
- mt7621 - JDCloud RE-SP-01B (mt7603 + mt7615)
- mt7621 - Cudy x6 v1 (mt7905 + mt7905)
- mt7621 - TP-Link EAP615-Wall (mt7905 + mt7975)
- mt7628 - Xiaomi Mi Router 4C (mt7603)

namiltd committed:
ramips: enable 6.12 testing kernel

The ramips now supports 6.12 kernel as testing.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18654
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

mcbridematt committed:
kernel: netdevices: rtl8366: ensure CONFIG_NET_DSA_REALTEK_RTL8366RB_LEDS is enabled

Add new setting CONFIG_NET_DSA_REALTEK_RTL8366RB_LEDS which
was introduced in kernel 6.12 and allows enabling LEDs for
the Realtek RTL8366RB subdriver.

Signed-off-by: Mathew McBride <matt@traverse.com.au>
Link: https://github.com/openwrt/openwrt/pull/18654
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

namiltd commented:
It builds correctly, you can test at your own risk.

![392169089-6eda64d3-5fa0-4882-be7c-28f726a4c621](https://github.com/user-attachments/assets/7a0d0cc3-36a4-4da6-924a-e6a49991d6df)

On Mercusys AC12G v1 (8M) (almost TP-Link Archer C5 V4 but without USB port) it starts up, but resets after half a minute from turning on the wifi and so on cyclically. Unfortunately, there is no information regarding the restart/crash on the serial terminal.

namiltd commented:
PR https://github.com/openwrt/openwrt/pull/18585 also works on 6.12.

everything411 commented:
flooding of `Trying to vfree() nonexistent vm area (c281b0ad)` errors in system log on mi router 4c when booting. seems related to module loading, after all modules loaded, these error stops
```
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.665271] ------------[ cut here ]------------
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.669965] WARNING: CPU: 0 PID: 78 at mm/vmalloc.c:3361 0x80119f9c
Thu May  1 23:38:31 2025 kern.err kernel: [   50.676359] Trying to vfree() nonexistent vm area (c281b0ad)
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.682141] Modules linked in: pppoe ppp_async nft_fib_inet nf_flow_table_inet pppox ppp_generic nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nft_quota nft_numgen nft_nat nft_masq nft_log nft_limit nft_hash nft_flow_offload nft_fib_ipv6 nft_fib_ipv4 nft_fib nft_ct nft_chain_nat nf_tables nf_nat nf_flow_table nf_conntrack slhc nfnetlink nf_reject_ipv6 nf_reject_ipv4 nf_log_syslog nf_defrag_ipv6 nf_defrag_ipv4 libcrc32c crc_ccitt sha512_generic seqiv sha3_generic jitterentropy_rng drbg hmac geniv rng cmac leds_gpio gpio_button_hotplug(O) crc32c_generic
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.734165] CPU: 0 UID: 0 PID: 78 Comm: kworker/0:2 Tainted: G        W  O       6.12.25 #0
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.742672] Tainted: [W]=WARN, [O]=OOT_MODULE
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.747094] Hardware name: Xiaomi Mi Router 4C
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.751605] Workqueue: events 0x80061788
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.755601] Stack : 804e4c78 00010000 809e4600 804af1a4 00000000 00000000 00000000 00000000
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.764137]         00000000 00000000 00000000 00000000 00000000 00000001 80a57cd8 9313d883
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.772672]         80a57d70 00000000 00000000 80a57bb8 00000038 80496be4 00000000 0000000b
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.781208]         00000000 ae94561f 57a74c49 00000000 804af1a4 00000000 804af1a4 00000000
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.789747]         80a57de8 00000d21 805f0000 80809305 00000000 80287008 00000000 80720000
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.798285]         ...
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.800775] Call Trace:
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.800794] [<80496be4>] 0x80496be4
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.806823] [<805f0000>] 0x805f0000
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.810373] [<80287008>] 0x80287008
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.813924] [<800061a8>] 0x800061a8
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.817475] [<800061b0>] 0x800061b0
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.821025] [<80476dc0>] 0x80476dc0
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.824575] [<80119f9c>] 0x80119f9c
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.828124] [<80119f9c>] 0x80119f9c
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.831674] [<8001fd70>] 0x8001fd70
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.835225] [<80119f9c>] 0x80119f9c
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.838775] [<80119f9c>] 0x80119f9c
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.842325] [<8001fe30>] 0x8001fe30
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.845911] [<80119f9c>] 0x80119f9c
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.849461] [<804990b0>] 0x804990b0
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.853067] [<800617dc>] 0x800617dc
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.856617] [<80031f94>] 0x80031f94
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.860195] [<800356d0>] 0x800356d0
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.863746] [<804990b0>] 0x804990b0
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.867296] [<80036984>] 0x80036984
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.870845] [<800412f0>] 0x800412f0
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.874404] [<80035c48>] 0x80035c48
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.877953] [<8049922c>] 0x8049922c
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.881504] [<80035aa8>] 0x80035aa8
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.885057] [<8003b38c>] 0x8003b38c
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.888607] [<8003b2b8>] 0x8003b2b8
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.892157] [<8003b2b8>] 0x8003b2b8
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.895708] [<80001b38>] 0x80001b38
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.899285]
Thu May  1 23:38:31 2025 kern.warn kernel: [   50.900849] ---[ end trace 0000000000000000 ]---
```

everything411 commented:
system reset occur immediately after enabling wireless on mi router 4c

neheb commented:
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/arch/mips/ralink?h=next-20250502&id=4f3089ad555601b494c91aa9dfbf9c3060cc3e73
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/arch/mips/ralink?h=next-20250502&id=9fd2ba4bf139bfe7781a24649d57133bd8af3624

namiltd commented:
Patch target/linux/generic/pending-6.12/811-pci_disable_usb_common_quirks.patch needs improvement.
On ramips/mt7621 it causes an error (https://github.com/namiltd/openwrt/actions/runs/14824787030/job/41616594627?pr=59):
```
Run make target/compile -j$(nproc) BUILD_LOG=1 || ret=$? .github/workflows/scripts/show_build_failures.sh
...
make[2]: Entering directory '/__w/openwrt/openwrt/openwrt/scripts/config'
make[2]: 'conf' is up to date.
make[2]: Leaving directory '/__w/openwrt/openwrt/openwrt/scripts/config'
 make[1] target/compile
 make[2] -C target/linux compile
    ERROR: target/linux failed to build.
make -r target/compile: build failed. Please re-run make with -j1 V=s or V=sc for a higher verbosity level to see what's going on
make: *** [/__w/openwrt/openwrt/openwrt/include/toplevel.mk:233: target/compile] Error 1
====== Make errors from logs/target/linux/compile.txt ======
  HDRINST usr/include/asm/unistd_n32.h
  INSTALL /__w/openwrt/openwrt/openwrt/build_dir/target-mipsel-openwrt-linux-musl_musl/linux-ramips_mt7621/linux-6.12.26/user_headers/include
make[4]: Leaving directory '/__w/openwrt/openwrt/openwrt/build_dir/target-mipsel-openwrt-linux-musl_musl/linux-ramips_mt7621/linux-6.12.26'
drivers/usb/host/pci-quirks.c:621:6: error: redefinition of 'usb_asmedia_modifyflowcontrol'
  621 | void usb_asmedia_modifyflowcontrol(struct pci_dev *pdev)
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from drivers/usb/host/pci-quirks.c:22:
drivers/usb/host/pci-quirks.h:49:20: note: previous definition of 'usb_asmedia_modifyflowcontrol' with type 'void(struct pci_dev *)'
   49 | static inline void usb_asmedia_modifyflowcontrol(struct pci_dev *pdev) {}
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[9]: *** [scripts/Makefile.build:229: drivers/usb/host/pci-quirks.o] Error 1
make[8]: *** [scripts/Makefile.build:478: drivers/usb/host] Error 2
make[7]: *** [scripts/Makefile.build:478: drivers/usb] Error 2
make[7]: *** Waiting for unfinished jobs....
make[6]: *** [scripts/Makefile.build:478: drivers] Error 2
make[5]: *** [/__w/openwrt/openwrt/openwrt/build_dir/target-mipsel-openwrt-linux-musl_musl/linux-ramips_mt7621/linux-6.12.26/Makefile:1949: .] Error 2
make[4]: *** [Makefile:224: __sub-make] Error 2
make[3]: *** [Makefile:24: /__w/openwrt/openwrt/openwrt/build_dir/target-mipsel-openwrt-linux-musl_musl/linux-ramips_mt7621/linux-6.12.26/.modules] Error 2
make[2]: *** [Makefile:12: compile] Error 2
time: target/linux/compile#3750.23#364.83#1055.97
Error: Process completed with exit code 2.
```

neheb commented:
does https://github.com/neheb/openwrt/commit/f21bb56de8291b8fcd672d241ca1ff09279ef52b work?

danpawlik commented:
similar to what @everything411  has:
```
[   37.518449] WARNING: CPU: 3 PID: 133 at mm/vmalloc.c:3361 vfree+0x1ac/0x2c4
[   37.525401] Trying to vfree() nonexistent vm area (0ac423ce)
[   37.531058] Modules linked in: pppoe ppp_async nft_fib_inet nf_flow_table_inet pppox ppp_generic nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nf
t_quota nft_numgen nft_nat nft_masq nft_log nft_limit nft_hash nft_flow_offload nft_fib_ipv6 nft_fib_ipv4 nft_fib nft_ct nft_chain_nat nf_tables nf_nat nf_flow_table nf_co
nntrack mt76x2e(O) mt76x2_common(O) mt76x02_lib(O) mt7603e(O) mt76(O) mac80211(O) cfg80211(O) slhc nfnetlink nf_reject_ipv6 nf_reject_ipv4 nf_log_syslog nf_defrag_ipv6 nf_
defrag_ipv4 libcrc32c crc_ccitt compat(O) cryptodev(O) crypto_user algif_skcipher algif_rng algif_hash algif_aead af_alg sha512_generic sha256_generic sha1_generic seqiv s
ha3_generic jitterentropy_rng drbg md5 crypto_hw_eip93 hmac geniv rng des_generic libdes cmac authencesn authenc leds_gpio gpio_button_hotplug(O) crc32c_generic
[   37.604657] CPU: 3 UID: 0 PID: 133 Comm: kworker/3:4 Tainted: G        W  O       6.12.25 #0
[   37.604681] Tainted: [W]=WARN, [O]=OOT_MODULE
[   37.604686] Hardware name: Xiaomi Mi Router 4A Gigabit Edition
[   37.604695] Workqueue: events do_free_init
[   37.604721] Stack : 80a00000 81c05cc4 00000003 8004d40c 81c05d34 81c05d03 00000003 ffffffff
[   37.604760]         800b3fd0 8180ab00 81809400 6e657665 00007374 00000001 81c05ce0 81878c40
[   37.604798]         00000000 00000000 808ff7e8 81c05af8 ffffefff 00000000 80a0cc80 00000ee2
[   37.604836]         81c05b44 00000ee4 80a0ccb0 ffffffff 00000001 00000000 808ff7e8 81c05df0
[   37.604873]         00000000 801e2970 80a05540 8102b600 00000000 00003c87 0000000c 80bd000c
[   37.604910]         ...
[   37.604919] Call Trace:
[   37.604923] [<80007ac4>] show_stack+0x28/0xf0
[   37.604948] [<807f7854>] dump_stack_lvl+0x70/0xb0
[   37.604962] [<8002c150>] __warn+0x9c/0x114
[   37.604983] [<8002c348>] warn_slowpath_fmt+0x180/0x188
[   37.605000] [<801e2970>] vfree+0x1ac/0x2c4
[   37.605017] [<800b4028>] do_free_init+0x58/0x84
[   37.605033] [<80048f34>] process_one_work+0x1cc/0x3c4
[   37.605052] [<80049d58>] worker_thread+0x320/0x494
[   37.605068] [<80053c98>] kthread+0x104/0x12c
[   37.605085] [<80002bd8>] ret_from_kernel_thread+0x14/0x1c
[   37.605100]
[   37.605104] ---[ end trace 0000000000000000 ]---
[   40.971438] mtk_soc_eth 1e100000.ethernet eth0: Link is Down
[   40.994442] mtk_soc_eth 1e100000.ethernet eth0: configuring for fixed/rgmii link mode
```

But the router starts normally, can work on it.

neheb commented:
Unlucky

KanjiMonster commented on target/linux/generic/hack-6.12/931-fix-compilation-warnings.patch:
```diff
@@ -0,0 +1,23 @@
+--- a/include/linux/irq.h
```
https://github.com/openwrt/openwrt/blob/main/target/linux/generic/PATCHES:

```
ALL patches must be in a way that they are potentially upstreamable, meaning:

- they must contain a proper subject
- they must contain a proper commit message explaining what they change
- they must contain a valid Signed-off-by line
```

Also include the actual warnings/errors in the commit message. Without them, it is hard to understand whether the suggested fix is the right approach.

KanjiMonster commented on target/linux/generic/hack-6.12/931-fix-compilation-warnings.patch:
```diff
@@ -0,0 +1,23 @@
+ 	lose_fpu_inatomic(1, prev);					\
+ 	if (tsk_used_math(next))					\
+-		__sanitize_fcr31(next);					\
++		{ __sanitize_fcr31(next); }				\
```
This has already been [fixed upstream since 6.12.2](https://github.com/gregkh/linux/commit/2a375beb4c3824bb984e65860ad7ec0197be91f7)

namiltd replied:
The history of this PR goes back to September 2024 and kernel 6.12, thanks for the info, I didn't notice it was patched in the meantime.

KanjiMonster commented on target/linux/ramips/patches-6.12/870-pci-rt2880-static-pcibios_init.patch:
```diff
@@ -0,0 +1,11 @@
+--- a/arch/mips/pci/pci-rt2880.c
```
https://github.com/openwrt/openwrt/blob/main/target/linux/generic/PATCHES:

```
ALL patches must be in a way that they are potentially upstreamable, meaning:

- they must contain a proper subject
- they must contain a proper commit message explaining what they change
- they must contain a valid Signed-off-by line
```

Also please send this upstream as AFAICT the issue is already in the kernel. That way we don't need to maintain yet another patch.

KanjiMonster replied:
That was true for everyone at one time. The general process is described in https://docs.kernel.org/process/submitting-patches.html, though it condenses to

1. Find out the correct git tree to base it on from MAINTAINERS
2. Make your changes and commit them
3. send the commit as a patch to the maintainers according to MAINTAINERS

And that's basically it. `./scripts/checkpatch.pl` will check your patch for formal (and some code) issues, `/scripts/get_maintainers.pl` to whom you should send the patches.

For 3, there is also a cli utility [b4](https://b4.docs.kernel.org/en/latest/contributor/overview.html) that you can use that even allows you to send out patches when your mail client/server can't send them out without breaking them (https://b4.docs.kernel.org/en/latest/contributor/send.html#authenticating-with-the-web-submission-endpoint).

namiltd replied:
Please someone do this for me.

KanjiMonster replied:
`pci-rt2880.c` is only used by the `ramips` target, so the patch should stay there. Generic already has far too many patches.

namiltd commented on target/linux/ramips/files/drivers/mmc/host/mtk-mmc/sd.c:
```diff
@@ -1965,14 +1965,16 @@ static irqreturn_t msdc_irq(int irq, void *dev_id)
 			if (intsts & MSDC_INT_ACMDCRCERR)
 				IRQ_MSG("XXX CMD<%d> MSDC_INT_ACMDCRCERR", cmd->opcode);
-			else
+			else {
```
Commit message updated

KanjiMonster replied:
The fix is still wrong, it only silences the warning, but it does not fix the actual issue:

The `IRQ_MSG()` macro needs to be fixed. Likely it needs its code to be wrapped/aliased to `do { } while (0)`.

See the fix for the fcr31 issue https://github.com/gregkh/linux/commit/2a375beb4c3824bb984e65860ad7ec0197be91f7 for an example.

namiltd replied:
Fixed

namiltd commented:
Problem discribed in https://github.com/openwrt/openwrt/pull/18654#issuecomment-2849484677 fixed with patch https://github.com/openwrt/openwrt/pull/18654/commits/d3c31ff8b54919e04b5113d1ce869144fb772fe2

namiltd commented:
@danpawlik,  @everything411 The problem should be fixed by the patch https://github.com/openwrt/openwrt/commit/be7ab6893a23631415e0462ce677b05471c61e62

danpawlik commented:
@namiltd can confirm! No traceback after recent changes. Thank you for your work

namiltd commented:
@everything411 Does WiFi work on your Mi Router 4c?

everything411 commented:
sadly, no. after enabling wireless, my router immediately stops responding and reboots in a few seconds.

everything411 commented:
I have been busy with my master's thesis over the past two months and I'm almost done. I will have some free time to test this branch on other devices, such as JDCloud RE-SP-01B mt7621 , in a few days.

namiltd commented:
Well, we have a riddle. Interestingly, on x86_64, WiFi client via backports/mac80211 works without any problems.

AzimsTech commented:
tested on ramips/mt7620 D-Link DIR-882 A1 no issues so far. 

```terminal
root@OpenWrt:~# ubus call system board
{
        "kernel": "6.12.26",
        "hostname": "DIR-882-STA",
        "system": "MediaTek MT7621 ver:1 eco:3",
        "model": "D-Link DIR-882 A1",
        "board_name": "dlink,dir-882-a1",
        "rootfs_type": "squashfs",
        "release": {
                "distribution": "OpenWrt",
                "version": "SNAPSHOT",
                "firmware_url": "https://downloads.openwrt.org/",
                "revision": "r29500-9e0ea9439e",
                "target": "ramips/mt7621",
                "description": "OpenWrt SNAPSHOT r29500-9e0ea9439e",
                "builddate": "1746704448"
        }
}
root@OpenWrt:~# uptime
01:43:38 up 1 day, 34 min,  load average: 0.08, 0.08, 0.06
 ```

everything411 commented:
Most likely. The wireless driver might operate the wireless hardware in an unsupported way and trigger a hardware reset without any traces. 

I noticed you're using the MT7620 platform, where the integrated 2.4GHz band uses the `rt2800soc` driver instead of `mt76`, while the 5GHz band uses the `mt76` driver.  

Mi Router 4C is based on the MT7628 platform, which also integrates 2.4GHz Wi-Fi. The integrated Wi-Fi shares its origins with MT7603, so it uses the `mt76` driver.  Mi Router 4C doesn't have a 5GHz band.

Can both the 2.4GHz and 5GHz bands trigger the crash on your device?

namiltd commented:
I turned on both networks at the same time so I don't know which one could be faulty. I currently do not have access to this router so I cannot test it.

namiltd committed:
kernel/ramips: Create kernel files for v6.12 (from v6.6)

This is an automatically generated commit.

When doing `git bisect`, consider `git bisect --skip`.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18654
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

namiltd committed:
kernel/ramips: Restore kernel files for v6.6

This is an automatically generated commit which aids following Kernel patch
history, as git will see the move and copy as a rename thus defeating the
purpose.

For the original discussion see:
https://lists.openwrt.org/pipermail/openwrt-devel/2023-October/041673.html

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18654
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

namiltd committed:
ramips: 6.12: remove obsolete patches

Remove obsolete patches already included in kernel 6.12.

Removed (already included):
 - 805-01-v6.9-pinctrl-Add-driver-for-Awinic-AW9523-B-I2C-GPIO-Expa.patch
 - 805-02-v6.9-pinctrl-aw9523-Add-proper-terminator.patch
 - 805-03-v6.10-pinctrl-aw9523-Destroy-mutex-on-remove.patch
 - 805-04-v6.10-pinctrl-aw9523-Use-correct-error-code-for-not-suppor.patch
 - 805-05-v6.10-pinctrl-aw9523-Always-try-both-ports-in-aw9523_gpio_.patch
 - 805-06-v6.10-pinctrl-aw9523-Make-use-of-struct-pinfunction-and-PI.patch
 - 805-07-v6.10-pinctrl-aw9523-Use-temporary-variable-for-HW-IRQ-num.patch
 - 805-08-v6.10-pinctrl-aw9523-Get-rid-of-redundant-U8_MAX-pieces.patch
 - 805-09-v6.10-pinctrl-aw9523-Remove-unused-irqchip-field-in-struct.patch
 - 805-10-v6.10-pinctrl-aw9523-Make-use-of-dev_err_probe.patch
 - 805-11-v6.10-pinctrl-aw9523-Sort-headers-and-group-pinctrl.patch
 - 805-12-v6.10-pinctrl-aw9523-Fix-indentation-in-a-few-places.patch

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18654
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

namiltd committed:
ramips: pci-rt2880: 6.12: change pcibios_init declaration

Fixes error:
arch/mips/pci/pci-rt2880.c:267:12: error: no previous prototype for 'pcibios_init' [-Werror=missing-prototypes]
  267 | int __init pcibios_init(void)
      |            ^~~~~~~~~~~~
cc1: all warnings being treated as errors
make[8]: *** [scripts/Makefile.build:229: arch/mips/pci/pci-rt2880.o] Error 1
make[7]: *** [scripts/Makefile.build:478: arch/mips/pci] Error 2

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18654
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

namiltd commented:
My router (TP-Link Archer C5 v4) crashes immediately after turning on 2.4GHz Wi-Fi (rt2800soc driver).

everything411 commented:
However, current openwrt snapshot builds are using mac80211 backported from kernel 6.12.6, and it works well on kernel 6.6. If mac80211 from 6.12 functions on 6.6, it’s unlikely that it would fail on 6.12 itself—that would be quite unexpected.

everything411 commented:
Kernel 6.12 works well on JDCloud RE-SB-01B, including both mt7603 2.4Ghz band and mt7615 5Ghz band.

everything411 commented:
the patch fails to apply?
```
$ git am mac80211-6.14.patch
Applying: mac80211: stage1
error: patch failed: package/kernel/mac80211/patches/ath9k/552-ath9k-ahb_of.patch:60
error: package/kernel/mac80211/patches/ath9k/552-ath9k-ahb_of.patch: patch does not apply
Patch failed at 0001 mac80211: stage1
hint: Use 'git am --show-current-patch=diff' to see the failed patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
$ git branch
* 6.12-ramips
```

namiltd commented:
Try on current updated main branch16 maj 2025 01:46 everything411 ***@***.***> napisał(a):
everything411 left a comment (openwrt/openwrt#18654)
the patch fails to apply?
$ git am mac80211-6.14.patch
Applying: mac80211: stage1
error: patch failed: package/kernel/mac80211/patches/ath9k/552-ath9k-ahb_of.patch:60
error: package/kernel/mac80211/patches/ath9k/552-ath9k-ahb_of.patch: patch does not apply
Patch failed at 0001 mac80211: stage1
hint: Use 'git am --show-current-patch=diff' to see the failed patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
$ git branch
* 6.12-ramips


—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

namiltd commented:
Try on current, refreshed branch.

keyodi commented:
Just wanted to share some feedback so far. 6.12 has been running great on ramips/mt7621 with mt7603 and mt7613. No issues no report over the last few days! Nice work!

namiltd commented:
Backports 6.14.5 works on x86. For ramips unfortunately I don't have a test router at the moment.
```
[   17.768140] Loading modules backported from Linux version v6.14.5-0-gaeaee199900e
[   17.786237] Backport generated by backports.git v6.1.110-1-43-g2e32643b84e8
[   17.816138] e1000e: Intel(R) PRO/1000 Network Driver
[   17.830031] e1000e: Copyright(c) 1999 - 2015 Intel Corporation.
[   18.162710] usb 2-1: reset high-speed USB device number 2 using ehci-pci
[   18.521627] mt7601u 2-1:1.0: ASIC revision: 76010001 MAC revision: 76010500
[   19.352419] mt7601u 2-1:1.0: EEPROM ver:0d fae:00
[   20.923930] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
[   20.924268] usbcore: registered new interface driver mt7601u
```

namiltd commented:
I tested backports 6.14.5 on ramips and unfortunately the restart problem did not go away. I am still looking for the cause.

keyodi commented:
> I tested backports 6.14.5 on ramips/mt7620 and unfortunately the restart problem did not go away. I am still looking for the cause.

Let me know if you would like me to test.

neheb commented on target/linux/generic/hack-6.12/931-mips-ralink-add-missing-include.patch:
```diff
@@ -0,0 +1,17 @@
+ 
+ #include <asm/irq_cpu.h>
+ #include <asm/mipsregs.h>
++#include <asm/time.h>
```
Why not backport the commits I linked to?

neheb replied:
hrm nvm. they're already present.

neheb replied:
Ah I see why upstream didn't hit this. It depends on CONFIG_IRQ_INTC, which is not the case for mt7621.

namiltd replied:
It was about another file.

everything411 commented:
As I mentioned earlier, since the mac80211 ported from 6.12 works on 6.6, it is unlikely to be problematic on 6.12. Also, since both `rt2800soc` and `mt76` are experiencing issues, it is highly likely that the problem does not lie within the WiFi NIC driver itself.
Affected devices are those with integrated wifi (via internal RBUS instead of PCIe). Could this issue possibly be related to the SoC itself, for example, SoC clock or power supply, etc.? We may need to examine the SoC code changes rather than the WiFi code.

namiltd committed:
mtk-mmc: replace empty body macros

Replace macros with empty body with 'do {} while (0)'.
Remove unnecessary semicolons.

Fixes:
drivers/mmc/host/mtk-mmc/sd.c: In function 'msdc_irq':
drivers/mmc/host/mtk-mmc/sd.c:1969:87: error: suggest braces around empty body in an 'else' statement [-Werror=empty-body]
 1969 |                                 IRQ_MSG("XXX CMD<%d> MSDC_INT_RSPCRCERR", cmd->opcode);
      |                                                                                       ^
drivers/mmc/host/mtk-mmc/sd.c:1975:84: error: suggest braces around empty body in an 'else' statement [-Werror=empty-body]
 1975 |                                 IRQ_MSG("XXX CMD<%d> MSDC_INT_CMDTMO", cmd->opcode);
      |                                                                                    ^
cc1: all warnings being treated as errors

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18654
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

namiltd committed:
mips: ralink: add missing include

Missing headers causes an error on kernel 6.12:
arch/mips/ralink/irq.c:86:5: error: no previous prototype for 'get_c0_perfcount_int' [-Werror=missing-prototypes]
   86 | int get_c0_perfcount_int(void)
      |     ^~~~~~~~~~~~~~~~~~~~
arch/mips/ralink/irq.c:92:14: error: no previous prototype for 'get_c0_compare_int' [-Werror=missing-prototypes]
   92 | unsigned int get_c0_compare_int(void)
      |              ^~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18654
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

KanjiMonster commented on target/linux/ramips/patches-6.12/931-mips-ralink-add-missing-include.patch:
```diff
@@ -0,0 +1,25 @@
+From: Mieczyslaw Nalewaj <namiltd@yahoo.com>
```
I checked, and like the pci-rt2800 issue this is an upstream issue and should be fixed there.

I'll gladly help you sending in patches upstream. I know it feels intimidating, but it isn't actually that complicated, especially for oneliner fixes.

namiltd commented:
I found https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=69708fbb2c698f262e03360d064c7066e0679953 which is present since kernel 6.7. I wonder if it can be the cause of crashes and why this patch is ​​not in patches-6.6.

neheb commented:
That's @DragonBluep 's patch.

LynxChaus commented:
build and installed on Xiaomi Mi Router 3G - all working (usb, ethernet, both wifi)

namiltd commented:
@DragonBluep Maybe you could take a closer look at the issue that causes integrated Wi-Fi devices to reset for kernel 6.12?

graysky2 commented:
I am using a ramips device as a switch, so no WiFi enabled. I can report functionality with this PR applied `r29749+10-d0d1f190683a`
```
Build system: x86/64
Build-tested: ramips/tplink_archer-a6-v3, x86/64
Run-tested: ramips/tplink_archer-a6-v3, x86/64
```

dangowrt commented:
Tested on Zbt WG3526 (16M) ((MT7621AT+MT7603E+MT7612E)
All works fine, Ethernet<->WiFi performance seems ok.

Regarding problems on MT7620:
> resets after half a minute from turning on the wifi and so on cyclically

If the timing is very predictable then I would guess it's a problem with the watchdog maybe...?

everything411 commented:
It should be noted that there are no serial console outputs—no kernel crash logs or any other diagnostic messages. This suggests that the system freeze occurs at a level where even the kernel's logging mechanism is no longer functional, possibly due to a hardware-level lockup concerning the SoC.

namiltd commented:
I suspected clocks, but the code is the same as in version 6.6 (including patches). I'll keep looking because I might have missed something.

namiltd committed:
ramips: 6.12: manually rebuilding patches

Manually rebuilding patches:
 - 001-v6.13-clocksource-drivers-ralink-Add-Ralink-System-Tick-Co.patch
 - 311-MIPS-use-set_mode-to-enable-disable-the-cevt-r4k-irq.patch
 - 314-MIPS-add-bootargs-override-property.patch
 - 405-mtd-spi-nor-Add-support-for-BoHong-bh25q128as.patch
 - 810-uvc-add-iPassion-iP2970-support.patch
 - 821-SPI-ralink-add-Ralink-SoC-spi-driver.patch
 - 825-i2c-MIPS-adds-ralink-I2C-driver.patch
 - 845-pwm-add-mediatek-support.patch
 - 860-ramips-add-eip93-driver.patch

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18654
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

namiltd committed:
ramips: 6.12: automatically rebuilding patches

Automatic rebasing of patches.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18654
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ramips: a workaround for system halt issue on 6.12 kernel

Fix potential invalid memory address access on ip_fast_csum().

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18654
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

namiltd commented:
The reboot issue has been fixed. It looks like the ramips target is ready to switch to the 6.12 kernel. 
Please approve this PR to expand the pool of testers.

namiltd commented:
Please test the functionality and compare the WiFi performance between v6.6 and 6.12.

everything411 commented:
Sorry, I'm currently on vacation and don't have access to the router with me. :)

everything411 commented:
I happen to have access to a Phicomm K2 (MT7620 + MT7612) and it runs well on version 6.12. As for performance, I'm not entirely sure—I don't know how this machine performed previously.

DragonBluep commented:
> I am asking ramips router users to test the functionality and compare WiFi performance between versions 6.6 and 6.12. On archer c5 v4 I found that WiFi performance dropped by 30%, Is this the effect of 0001-ramips-a-workaround-for-system-halt-issue-on-6.12-ke.patch? I don't know. @DragonBluep Could changes to the ip_fast_csum function be causing this issue 

No. Unlikely.

> or does this change just mitigate a problem that is elsewhere?

I think something wrong in the net framework. If someone have time && related knowledge, please track the function callback to check what's wrong here.

> The reboot issue has been fixed. It looks like the ramips target is ready to switch to the 6.12 kernel. Please approve this PR to expand the pool of testers.

The root cause has not been found yet, my patch is just a workaround.

namiltd commented:
I still don't understand why the problem will occur for SoC and not for additional chips by PCIe.

DragonBluep commented on target/linux/ramips/mt7621/config-6.12:
```diff
@@ -308,6 +314,7 @@ CONFIG_WATCHDOG_CORE=y
 CONFIG_WEAK_ORDERING=y
 CONFIG_XPS=y
 CONFIG_XXHASH=y
+CONFIG_XZ_DEC_BCJ=y
```
These configs need to be refreshed again.

keyodi commented:
> I am asking ramips router users to test the functionality and compare WiFi performance between versions 6.6 and 6.12.

I haven't noticed any performance differences between 6.6 -> 6.12 on eth or wifi on (6.12-ramips). I do have packet steering set to All Cpus and enabled hardware offload. Hitting expected bandwidth.

namiltd commented:
WiFi performance degradation will most likely only occur with the MT7620 (2.4GHz) and MT7628 chips.

everything411 commented:
The reboot issue on mt7620 is still there for 6.12 kernel. https://github.com/openwrt/openwrt/pull/17561

namiltd commented:
PR #17561 is about sporadic restarts, so a different issue

leobsky commented:
Tested on cudy x6 v1 (mt7621) no issues, same performance than kernel 6.6.

everything411 commented:
> PR #17561 is about sporadic restarts, so a different issue. It also occurs for kernel 6.6.

Yes I mean the sporadic restarts still occur on 6.12. The boot loop concerning 2.4Ghz wireless we talked about earlier in this issue has gone after applying DragonBluep's patch. However, there are still annoying sporadic restarts on mt7620 machines with 6.12 kernel, which is most likely the same issue as kernel 6.6.

namiltd committed:
kernel/ramips: update config for kernel 6.12

Update target ramips config for kernel 6.12.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18654
[Remove some more option already in generic config]
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

namiltd commented:
@everything411 I recommend https://github.com/x-wrt/x-wrt/commit/f21667c722d86648b79681c56efc9d374193dc17.
Prevents restarting at an earlier stage of loading.

namiltd commented:
Call Trace of error occurrence ihl==0:
```
[  174.162908] [<804e04f4>] ip_send_check+0xdc/0xe4
[  174.167642] [<80538a68>] inet_gso_segment+0x1fc/0x3b8
[  174.172830] [<80488ec8>] skb_mac_gso_segment+0xb0/0x118
[  174.178181] [<80488fb0>] __skb_gso_segment+0x80/0x1ac
[  174.183345] [<80446e1c>] validate_xmit_skb.isra.0+0x1ac/0x3bc
[  174.189228] [<80447838>] __dev_queue_xmit+0x1ac/0xbf0
[  174.194394] [<804e11f4>] ip_finish_output2+0x40c/0x714
[  174.199656] [<804dcc74>] ip_sublist_rcv+0x220/0x26c
[  174.204647] [<804dd3d0>] ip_list_rcv+0xfc/0x158
[  174.209285] [<80449e50>] __netif_receive_skb_list_core+0x180/0x224
[  174.215604] [<8044a3b8>] netif_receive_skb_list_internal+0x1d0/0x2dc
[  174.222100] [<8044aa54>] napi_complete_done+0x68/0x2a4
[  174.227355] [<80401588>] fe_poll+0x52c/0x73c
[  174.231743] [<8044accc>] __napi_poll+0x3c/0x1c4
[  174.236377] [<8044b1f4>] net_rx_action+0x144/0x2b0
[  174.241278] [<80028b7c>] handle_softirqs+0x108/0x288
[  174.246362] [<80002178>] except_vec_vi_end+0xc4/0xd0
```
I'm trying to find a reason.

everything411 commented:
Confirmed working on Mi router 4C. No obvious WiFi performance degradation.

namiltd commented:
I don't know any, so the only thing I can do in such situations is to display the contents of variables in the log.

keyodi commented:
Out of curiosity, do you still receive the error with GRO/GSO/GRO_Frag disabled?

namiltd commented:
Ping @luizluca

hauke commented on target/linux/ramips/patches-6.12/900-pci-rt2880-static-pcibios_init.patch:
Please send this patch upstream too. Maybe add a stable tag too and a fixes reference when the WError was introduced.

namiltd replied:
I don't send patches to upstream, I'm scared by the complexity of its structure. So please, someone do it for me if possible.

hauke commented on target/linux/ramips/mt7621/config-6.12:
```diff
@@ -120,6 +122,8 @@ CONFIG_MDIO_BUS=y
 CONFIG_MDIO_DEVICE=y
 CONFIG_MDIO_DEVRES=y
 CONFIG_MEDIATEK_GE_PHY=y
+# CONFIG_MFD_88PM886_PMIC is not set
+# CONFIG_MFD_ROHM_BD96801 is not set
```
This is both part of `target/linux/generic/config-6.12` no need to add it here.

hauke commented on target/linux/ramips/mt7621/config-6.12:
```diff
@@ -101,6 +102,7 @@ CONFIG_I2C=y
 CONFIG_I2C_ALGOBIT=y
 CONFIG_I2C_BOARDINFO=y
 CONFIG_I2C_CHARDEV=y
+# CONFIG_I2C_DESIGNWARE_CORE is not set
```
applied it before merge
